File: homework1/tests/warmup-test.js
const {
  change, stripQuotes, scramble, say, powers, interleave,
  powersGenerator, cylinder, makeCryptoFunctions, randomName,
} = require('../src/warmup');

require('should'); // eslint-disable-line import/no-extraneous-dependencies


// Helper for the scramble tests
function anagramsOfEachOther(s, t) {
  return s.split('').sort().join('') === t.split('').sort().join('');
}

// Helper for testing the callbacky problems
function generatorToArray(generator, ...args) {
  const result = [];
  generator(...args, x => result.push(x));
  return result;
}

describe('change', () => {
  it('handles zero', () => {
    change(0).should.eql([0, 0, 0, 0]);
  });

  it('computes answers for small integer values fine', () => {
    change(97).should.eql([3, 2, 0, 2]);
    change(8).should.eql([0, 0, 1, 3]);
    change(250).should.eql([10, 0, 0, 0]);
    change(144).should.eql([5, 1, 1, 4]);
    change(97).should.eql([3, 2, 0, 2]);
  });

  it('handles large values', () => {
    // This test only passes if the solution is efficient
    change(100000000000).should.eql([4000000000, 0, 0, 0]);
  });

  it('throws the proper exception for negative arguments', () => {
    (() => change(-50)).should.throw(RangeError);
  });
});

describe('stripQuotes', () => {
  it('works on the empty string', () => {
    stripQuotes('').should.eql('');
  });

  it('strips quotes properly for non-empty strings', () => {
    stripQuotes('Hello, world').should.eql('Hello, world');
    stripQuotes('"\'').should.eql('');
    stripQuotes('a"""\'\'"z').should.eql('az');
  });
});

describe('scramble', () => {
  it('scrambles properly', () => {
    ['a', 'rat', 'JavaScript testing', '', 'zzz', '^*^*)^▱ÄÈËɡɳɷ'].forEach(s =>
      anagramsOfEachOther(s, scramble(s)).should.be.true);
  });

  it('is really random (produces all permutations)', () => {
    const possibilities = new Set('ABC ACB BAC BCA CAB CBA'.split(' '));
    for (let i = 0; i < 200; i += 1) {
      possibilities.delete(scramble('ABC'));
    }
    possibilities.size.should.eql(0);
  });
});

describe('say', () => {
  it('works when there are no words', () => {
    say().should.eql('');
  });

  it('works when there are words', () => {
    say('hi')().should.eql('hi');
    say('hi')('there')().should.eql('hi there');
    say('hello')('my')('name')('is')('Colette')().should.eql('hello my name is Colette');
  });
});

describe('powers', () => {
  it('generates sequences of powers properly', () => {
    generatorToArray(powers, 2, -5).should.eql([]);
    generatorToArray(powers, 7, 0).should.eql([]);
    generatorToArray(powers, 3, 1).should.eql([1]);
    generatorToArray(powers, 2, 63).should.eql([1, 2, 4, 8, 16, 32]);
    generatorToArray(powers, 2, 64).should.eql([1, 2, 4, 8, 16, 32, 64]);
  });
});

describe('The powers generator', () => {
  it('works as expected', () => {
    const g1 = powersGenerator(2, 1);
    g1.next().should.eql({ value: 1, done: false });
    g1.next().should.eql({ value: undefined, done: true });
    const g2 = powersGenerator(3, 100);
    g2.next().should.eql({ value: 1, done: false });
    g2.next().should.eql({ value: 3, done: false });
    g2.next().should.eql({ value: 9, done: false });
    g2.next().should.eql({ value: 27, done: false });
    g2.next().should.eql({ value: 81, done: false });
    g2.next().should.eql({ value: undefined, done: true });
  });
});

describe('interleave', () => {
  it('interleaves properly', () => {
    interleave([]).should.eql([]);
    interleave([1, 4, 6]).should.eql([1, 4, 6]);
    interleave([], 2, 3).should.eql([2, 3]);
    interleave([1], 9).should.eql([1, 9]);
    interleave([8, 8, 3, 9], 1).should.eql([8, 1, 8, 3, 9]);
    interleave([2], 7, '8', {}).should.eql([2, 7, '8', {}]);
  });
});

describe('The cylinder function', () => {
  it('makes a cylinder from both arguments', () => {
    const c = cylinder({ radius: 10, height: 5 });
    c.height.should.eql(5);
    c.radius.should.eql(10);
  });
  it('defaults the radius to 1', () => {
    const c = cylinder({ height: 5 });
    c.height.should.eql(5);
    c.radius.should.eql(1);
  });
  it('defaults the height to 1', () => {
    const c = cylinder({ radius: 5 });
    c.height.should.eql(1);
    c.radius.should.eql(5);
  });
  it('accepts an empty object', () => {
    const c = cylinder({});
    c.height.should.eql(1);
    c.radius.should.eql(1);
  });
  it('computes volumes and surface areas correctly', () => {
    const c = cylinder({ radius: 2, height: 10 });
    c.volume().should.be.approximately(40 * Math.PI, 0.000001);
    c.surfaceArea().should.be.approximately(48 * Math.PI, 0.000001);
  });
  it('mutates with stretch and widen', () => {
    const c = cylinder({ radius: 2, height: 10 });
    c.widen(3);
    c.radius.should.eql(6);
    c.stretch(2);
    c.height.should.eql(20);
  });
  it('changes volumes after stretch and widen', () => {
    const c = cylinder({ radius: 2, height: 10 });
    c.widen(3);
    c.volume().should.be.approximately(360 * Math.PI, 0.000001);
    c.stretch(2);
    c.volume().should.be.approximately(720 * Math.PI, 0.000001);
  });
  it('has an immutable radius field', () => {
    const c = cylinder({ radius: 2, height: 10 });
    c.radius = 100;
    c.radius.should.eql(2);
  });
  it('has an immutable height field', () => {
    const c = cylinder({ radius: 2, height: 10 });
    c.height = 100;
    c.height.should.eql(10);
  });
});

describe('crypto functions', () => {
  it('encrypt and decrypt okay', () => {
    const [e, d] = makeCryptoFunctions('super dog', 'aes-256-cbc');
    e('Hello world').should.eql('9714236cbedfd8d9799acea4ea79e6fe');
    d('9714236cbedfd8d9799acea4ea79e6fe').should.eql('Hello world');
    ['', 'abc', 'zπøj#•¶å≈’’'].forEach(s => d(e(s)).should.eql(s));
  });
  it('throws an error for an unknown algorithm', () => {
    (() => makeCryptoFunctions('super dog', 'asdf*')[0]('Hello')).should.throw(Error);
  });
});

describe('The random name function', () => {
  it('returns a promise, anyway', () => {
    const p = randomName({ gender: 'female', region: 'canada' });
    ('then' in p).should.be.ok();
  });
  it('produces a resolved promise with a name and surname', (done) => {
    randomName({ gender: 'female', region: 'canada' }).then((name) => {
      // You might try: console.log(name); // eslint-disable-line no-console
      name.length.should.be.above(3);
      name.should.containEql(' ');
    }).then(done, done);
  });
  it('produces a rejected promise for an unknown gender', (done) => {
    randomName({ gender: 'fefwefemale', region: 'canada' }).catch((error) => {
      error.message.should.containEql('400');
    }).then(done, done);
  });
});

File: homework2/tests/warmup_test.py
import re
import math
import pytest
from warmup import (change, strip_quotes, scramble, say, triples, powers,
                    interleave, Cylinder, make_crypto_functions, random_name)


def test_change():
    assert change(0) == (0, 0, 0, 0)
    assert change(97) == (3, 2, 0, 2)
    assert change(8) == (0, 0, 1, 3)
    assert change(250) == (10, 0, 0, 0)
    assert change(144) == (5, 1, 1, 4)
    assert change(97) == (3, 2, 0, 2)
    assert change(100000000000) == (4000000000, 0, 0, 0)
    with pytest.raises(ValueError) as excinfo:
        change(-50)
    assert str(excinfo.value) == 'amount cannot be negative'

def test_strip_quotes():
    assert strip_quotes('') == ''
    assert strip_quotes('Hello, world') == 'Hello, world'
    assert strip_quotes('"\'') == ''
    assert strip_quotes('a"""\'\'"z') == 'az'

def test_scramble():
    for s in ['a', 'rat', 'JavaScript testing', '', 'zzz', '^*))^*>^▱ÄÈËɡɳɷ']:
        assert sorted(s) == sorted(scramble(s))
    possibilities = set(['ABC', 'ACB', 'BAC', 'BCA', 'CAB', 'CBA'])
    for _ in range(200):
        possibilities.discard(scramble('ABC'))
    assert not possibilities

def test_say():
    assert say() == ''
    assert say('hi')() == 'hi'
    assert say('hi')('there')() == 'hi there'
    assert say('hello')('my')('name')('is')('Colette')() == 'hello my name is Colette'

def test_triples():
    assert triples(0) == []
    assert triples(5) == [(3, 4, 5)]
    assert set(triples(40)) == set([(3, 4, 5), (5, 12, 13), (6, 8, 10), (7, 24, 25), (8, 15, 17),
                                    (9, 12, 15), (10, 24, 26), (12, 16, 20), (12, 35, 37),
                                    (15, 20, 25), (15, 36, 39), (16, 30, 34), (18, 24, 30),
                                    (20, 21, 29), (21, 28, 35), (24, 32, 40)])

def test_powers():
    p = powers(2, 10)
    assert next(p) == 1
    assert next(p) == 2
    assert next(p) == 4
    assert next(p) == 8
    with pytest.raises(StopIteration):
        next(p)
    assert list(powers(2, -5)) == []
    assert list(powers(7, 0)) == []
    assert list(powers(3, 1)) == [1]
    assert list(powers(2, 63)) == [1, 2, 4, 8, 16, 32]
    assert list(powers(2, 64)) == [1, 2, 4, 8, 16, 32, 64]

def test_interleave():
    assert interleave([]) == []
    assert interleave([1, 4, 6]) == [1, 4, 6]
    assert interleave([], 2, 3) == [2, 3]
    assert interleave([1], 9) == [1, 9]
    assert interleave([8, 8, 3, 9], 1) == [8, 1, 8, 3, 9]
    assert interleave([2], 7, '8', {}) == [2, 7, '8', {}]
    a = [1, 2, 3, 4]
    assert interleave(a, 10, 20, 30) == [1, 10, 2, 20, 3, 30, 4]
    # Test input list not destroyed
    assert a == [1, 2, 3, 4]

def test_cylinder():
    c = Cylinder(radius=10, height=5)
    assert c.height == 5
    assert c.radius == 10
    c = Cylinder(height=5)
    assert c.height == 5
    assert c.radius == 1
    c = Cylinder(radius=5)
    assert c.height == 1
    assert c.radius == 5
    c = Cylinder()
    assert c.height == 1
    assert c.radius == 1
    c = Cylinder(radius=2, height=10)
    assert pytest.approx(c.volume, 0.000001) == 40 * math.pi
    assert pytest.approx(c.surface_area, 0.000001) == 48 * math.pi
    c.widen(3)
    assert c.radius == 6
    c.stretch(2)
    assert c.height == 20
    assert pytest.approx(c.surface_area, 0.000001) == 312 * math.pi
    assert pytest.approx(c.volume, 0.000001) == 720 * math.pi

def test_crypto():
    assert isinstance(make_crypto_functions(b"A2qK5XG3qX1MfLrGacD9AGVG2sbZYkvFFki94qbkVhE="), tuple)
    e, d = make_crypto_functions(b"A2qK5XG3qX1MfLrGacD9AGVG2sbZYkvFFki94qbkVhE=")
    for s in [b'', b'\xfe9iP\x05\x22\x490opXZ@1##', b'Helllllllllooooooo world']:
        assert d(e(s)) == s

def test_random_name():
    p = random_name(gender='female', region='canada')
    assert isinstance(p, str)
    assert len(p) > 3
    assert ', ' in p
    with pytest.raises(ValueError) as excinfo:
        random_name(gender='fjweiuw', region='canada')
    assert re.match(r'{"error":\s*"Invalid gender"}', str(excinfo.value))
